lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -0,0 +1,1617 @@
+      auto i = bb.begin(), e = bb.end();
+      while (i != e) {
+        SILInstruction *I = &*i;
+        
+        if (auto FRI = dyn_cast<FunctionRefInst>(I)) {
+          if (auto Fn = FRI->getReferencedFunctionOrNull()) {
+            if (auto DC = Fn->getDeclContext()) {
+              if (auto AD = dyn_cast_or_null<AccessorDecl>(DC)) {
+                if (AD->getAccessorKind() == AccessorKind::Get) {
+                  GetterInst = FRI;
+                  return true;
+                }
+              }
+            }
+          }
+        }
+        
+        if (auto CMI = dyn_cast<ClassMethodInst>(I)) {
+          if (auto Member = CMI->getMember()) {
+            if (auto AD = dyn_cast<AccessorDecl>(Member.getDecl())) {
+              if (AD->getAccessorKind() == AccessorKind::Get) {
+                GetterInst = CMI;
+                return true;
+              }
+            }
+          }
+        }
+        
+        ++i;
+      }
+    }
+    return false;
+  }
+};
+
+//===----------------------------------------------------------------------===//
+//                         MARK: Traverser Interface
+//===----------------------------------------------------------------------===//
+
+class VarUsageInfo {
+  
+public:
+
+  SILInstruction *EntryInst;
+  
+  SILInstruction *MisusedGetter = nullptr;
+  bool DidRead = false;
+  bool DidModify = false;
+  bool DidDeferInit = false;
+  bool IsWeakCapture = false;
+  VarDecl *InitialReference = nullptr;
+  
+  VarUsageInfo(SILInstruction *EntryInst) : EntryInst(EntryInst) {
+    collectUsageInfo();
+  }
+  
+  void addEntryInst(SILInstruction *I) {
+    EntryInst = I;
+    collectUsageInfo();
+  }
+  
+private:
+  
+  void searchSetterParam() {
+    auto VRT = VariableReadTraverser(EntryInst);
+    if (!VRT.didRead()) {
+      if (auto Fn = EntryInst->getFunction()) {
+        auto SRGT = SetterReadsGetterTraverser(Fn);
+        if (SRGT.findGetterAccess()) {
+          MisusedGetter = SRGT.GetterInst;
+        }
+      }
+    }
+    DidRead = true; // Set this to mute other diagnostics
+  }
+  
+  /// Get the first \c DebugValueInst to reference the value if it is not
+  /// the entry instruction
+  VarDecl *getInitialReference() {
+    auto DVI = dyn_cast<DebugValueInst>(EntryInst);
+    if (!DVI) {
+      return nullptr;
+    }
+    
+    if (auto DebugVD = getVarDecl(EntryInst)) {
+      if (DebugVD->isCaptureList()) {
+        return nullptr;
+      }
+      if (auto ParentStmt = DebugVD->getRecursiveParentPatternStmt()) {
+        if (isa<CaseStmt>(ParentStmt)) {
+          return nullptr;
+        }
+      }
+    }
+    
+    auto ReferencedValue = DVI->getAllOperands()[0].get();
+    
+    DebugValueInst *InitialRefInst = DVI;
+    for (auto Use : ReferencedValue->getUses()) {
+      auto UserDVI = dyn_cast<DebugValueInst>(Use->getUser()); 
+      if (!DVI) {
+        continue;
+      }
+      if (instructionPrecedes(UserDVI, DVI, EntryInst->getParent())) {
```

Comment:
It looks like there will be quadratic behavior here in any pathological cases, because there is a loop inside `instructionPrecedes`. I don't have a good sense of what code will trigger this worst-case behavior though...

Reply:
I think I may be able to re-implement this without looping again. I may be able to store a worklist of seen instructions and check if the instruction has been seen before the current instruction. I'm not completely sure this will work, but I will take a look and try to optimize this, or at least document which cases will trigger worst-case behavior.

Reply:
Yeah, I think it would be useful to at least first have a sense of if this is a problem in practice (for example, you could log comma-separated values to stderr when compiling some code, import the csv file into a spreadsheet program and make a box plot) before going ahead to optimize this.

Reply:
Great idea, thanks!

Reply:
I removed this loop, and instead compare to a set of seen instructions. Should reduce time complexity.